pkg/autoscaler/autoscaler.go

```diff
@@ -361,10 +369,106 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (int32, bool) {
 	return desiredPodCount, true
 }
 
+func (a *Autoscaler) aggregateData(now time.Time, stableWindow, panicWindow time.Duration) (*totalAggregation, *totalAggregation, map[string]Stat) {
+	a.statsMutex.Lock()
+	defer a.statsMutex.Unlock()
+
+	// 60 second window
+	stableData := newTotalAggregation(stableWindow)
+
+	// 6 second window
+	panicData := newTotalAggregation(panicWindow)
+
+	// Last stat per Pod
+	lastStat := make(map[string]Stat)
+
+	// Accumulate stats into their respective buckets
+	for key, stat := range a.stats {
+		instant := key.time
+		if instant.Add(panicWindow).After(now) {
+			panicData.aggregate(stat)
+		}
+		if instant.Add(stableWindow).After(now) {
+			stableData.aggregate(stat)
+
+			// If there's no last stat for this pod, set it
+			if _, ok := lastStat[stat.PodName]; !ok {
+				lastStat[stat.PodName] = stat
+			} else if lastStat[stat.PodName].Time.Before(*stat.Time) {
+				// If the current last stat is older than the new one, override
+				lastStat[stat.PodName] = stat
+			}
+		} else {
+			// Drop metrics after 60 seconds
+			delete(a.stats, key)
+		}
+	}
+	return stableData, panicData, lastStat
+}
+
+func (a *Autoscaler) targetConcurrency() float64 {
+	a.targetMutex.RLock()
+	defer a.targetMutex.RUnlock()
+	return a.target
+}
+
 func (a *Autoscaler) podCountLimited(desiredPodCount, currentPodCount float64) float64 {
 	return math.Min(desiredPodCount, a.Current().MaxScaleUpRate*currentPodCount)
 }
 
+func (a *Autoscaler) readyPods(now time.Time) (float64, error) {
+	if v, ok := a.readyPodsFromCache(now); ok {
+		return v, nil
+	}
+
+	return a.readyPodsFromLister(now)
```

Comment:
Do we need to make sure that this is only called once per rounded `now`? I'm not sure it's worth it given the cached notion of the endpointInformer but it might be worth thinking about.

We could solve that by making the lock in `readyPodsFromLister` span the whole function again and check for availability in the cache again directly after acquiring the lock.

Does that make sense @vagababov?

Reply:
Holding lock over the RPC, is something I'd prefer not to do. Since the informer is backed by a cache, I think the penalty of additional Endpoints() call is probably negligible (though same applies to the time the lock is held, unless there's a cache miss). 

A semi-useless experiment would be to add a metric that reports the number of times `cache[now]` was already set. And then see what's the value in real life (and in addition measure latency of the `readypodsfromlister` call), but I am not sure if we care that much.

I'd rather leave it as is, since it's simpler than
```
checkCache
lock
checkCache
checkLister
updateCache
unlock
```

It's like creating singletons in Java all over again :-)

Reply:
To be clear, the current behavior is
```
checkCache
checkLister
lock
updateCache
unlock
```
@markusthoemmes proposes
```
lock
checkCache
checkLister
updateCache
unlock
```
so that `checkLister` could be only called once per rounded `now`, right?

I agree with @vagababov since the informer is backed by a cache.

Reply:
No I actually proposed @vagababov's double-check behavior akin to Java Singletons yeah. Premature optimization, I'm fine keeping it as is.